8.0 Import regular PDF invoices (via invoice2data), generate and import UBL and ZUGFeRD invoices

alexis-via opened the pull request.

alexis-via committed:
Add module account_invoice_pdf_import

alexis-via committed:
Support settings price_unit with a price_include tax

alexis-via committed:
Use lib invoice2data

alexis-via committed:
Add ability to force invoice line label

Fix for multi-company (field on res.partner should be a property)
Small usability improvements

alexis-via committed:
Add installation section

alexis-via committed:
Better error handling

alexis-via committed:
Add ability to force tax amount

alexis-via committed:
Match partner even if the VAT number has spaces in Odoo

alexis-via committed:
Add support for ZUGFeRD

alexis-via committed:
Add support for IBAN + BIC

Add help message inside wizard
Block if invoice already exists

alexis-via committed:
Add YAML test and demo data

alexis-via committed:
Add support for updating an existing invoice

alexis-via committed:
Add links in wizard help message

alexis-via committed:
YAML test for ZUGFeRD

Add YAML test for invoice update
Fallback on matching on partner name

alexis-via committed:
Comment test code

alexis-via committed:
FIX import of refunds

Code cleanup
Finish the TODO list !

alexis-via committed:
More info in README

alexis-via committed:
FIX README

alexis-via committed:
FIX an error in my previous commit

alexis-via committed:
Add an idea for the roadmap

alexis-via committed:
Fix RST syntax

alexis-via committed:
Try to fix RST syntax

alexis-via committed:
FIX second code block

alexis-via commented:
This PR is Akretion's Christmas present to the Odoo community ! You can read the story behind the development of this module with some screenshots here: http://www.akretion.com/blog/akretions-christmas-present-for-the-odoo-community 

We hope that you will like our present ! :) The README gives detailed instructions on how to install, configure and use the module.

alexis-via committed:
Add FR translation. Enhance some strings.

When Odoo propose to update an existing invoice, filter on the relevant invoice type

alexis-via committed:
Update external_dependencies

alexis-via commented:
Travis is in fact green (the red comes from a bad Transiflex password and some missing modules error, but unrelated to this module).

alexis-via committed:
Add link to blog post in README

alexis-via commented:
FYI, I'm currently thinking about renaming the module from "account_invoice_pdf_import" to "account_invoice_import", because I would like to be able to import PDF files or XML files. I'm looking at how the invoicing system of the suppliers of the French administration works (Chorus), and it seems that they accept direct XML import, and not just PDF. So it may become a normal practise for a few "special" suppliers/customers to send XML files instead of PDF, and it would be better to handle both with the same module.

It's the time to take the good decision on the module name, because it's a pain to change the name of the module later on.

alexis-via committed:
FIX default value

sbidoul commented:
@alexis-via you could do a base module with addons modules to handle specific formats. Similar to bank_statement_import & co...

alexis-via commented:
@sbidoul yes, it's a good idea. But this PR is for the "base" module, and my question is: should I remove 'pdf' from the name of this base module or not. I think I'll remove it and the name will be "account_invoice_import" because it is more generic. This module would support PDF and XML (CII and UBL), and other non-standard or exotic formats would be supported by additionnal modules that extend this module.

sbidoul commented:
+1 to remove pdf from the name.

alexis-via committed:
Update invoice2data library URL, now that my PR has been merged

alexis-via committed:
Rename module from account_invoice_pdf_import to account_invoice_import

alexis-via committed:
FIX raise error when we don't identify a partner

brother-bernard commented:
:+1: using this code in production

alexis-via committed:
Update README

alexis-via committed:
Update FR translation

MichelGuenard commented:
I am an Odoo Saas online user V8 and I am NOT sure if I can use this new module, which will highly ease my work as a lazzy accountant! More seriously this will fill partially the gap of no-EDI modules available with Odoo.
In case my online database allows to use such module, could someone tell me how to achieve this because I am totally new with Github and I do not understand how to download the module in order to import it with my Odoo system.
Many thanks to Akretion for this contribution

elicoidal commented:
I am sorry but OCA modules are not available in the official SaaS offer 
from Odoo.
You will need to find a hoster that includes OCA modules or get a server 
by yourself.

MichelGuenard commented:
Thanks for this confirmation. I will consider a new system allowing the OCA modules.

alexis-via committed:
Add instructions to use my PPA for poppler-utils on Ubuntu 14.04 LTS

alexis-via committed:
Update README with instructions on OCR

alexis-via committed:
Add support for direct import of XML files (YAML test provided)

Rename test/pdf/ to test/invoices/ to host XML invoices too

alexis-via committed:
Fix pylink

Commit:
Sometimes, it's easier to extract the total tax amount instead of the total untaxed (depends on invoice layout)

alexis-via committed:
Add an explaination about date parsing in invoice2data

Commit:
Add optional support for start/end dates

alexis-via committed:
Adapt to latest changes in invoice2data

PEP8 stuff

alexis-via committed:
Only change name_get of invoice for our use-case, not everywhere

alexis-via committed:
Update README with latest changes in invoice2data lib

sebnege commented:
+1: it works well, using it in production

AlexandreFournier commented:
+1

cdubuit commented:
:+1: using this PR in production

alexis-via committed:
Update with the latest changes in invoice2data lib

Commit:
Add support for partner_email to match partners

PEP8 fix

alexis-via committed:
Update installation procedure.

alexis-via committed:
Add modules base_zugferd, account_invoice_import_zugferd and account_invoice_zugferd (see README.rst for more info)

Move the ZUGFeRD/CII-specific code of account_invoice_import to account_invoice_import_zugferd

alexis-via committed:
FIX avoid a crash on draft invoice without date

alexis-via committed:
Update some README

alexis-via committed:
Update travis config

alexis-via committed:
Update README

OSguard commented:
Should ZUGFeRD be part of OCA/l18n-germany ?

alexis-via commented:
@OSguard ZUGFeRD is such a great standard that it should not be limited to german companies ! As you have seen, the implementation has been developped by a French company and is currently used when invoicing between French companies (Akretion and its French customers) :)

More seriously, that's a good question. I know that the standard has been written by a German standardisation body, but I read that they don't only target Germany but they would like to see this standard used internationally (I'll try to find that reference again). On my side, I would like to see this great standard used through-out Europe and also internationally.

OSguard commented:
It is developed by a a german ministerie and compatible with eu laws. They hand the draft over to make it standard to EU but the process is ongoing. I don't know if there other counter concepts to become standard.
So it is law in Germany, you can may use it in other countries as well.
So OCA/l18n-eu do not exist. But the important part that we have a free implementation inside the OCA where it located just to have clear rules.

alexis-via commented:
@OSguard I think that this ZUGFeRD standard should be used by anybody from any country who want to exchange electronic invoices. So, hiding this module inside l10n-xxx is probably not the best way to give visibility to this great feature. Electronic invoices are the future of invoices, not just for Germany or Europe, but to the whole world ! And ZUGFeRD is based on international formats such as Cross Industry Invoice (CII) of UN/CEFACT, which is designed for the whole world. I must say that I hesitated in naming the module account_invoice_zugferd ; I could have named it account_invoice_cii, but the embedding of the XML inside the PDF has been specified by ZUGFeRD so I decided to name it zugferd...

alexis-via committed:
Parsing of ZUGFeRD PDF invoice should now work when there are several attachments

alexis-via committed:
Log raw result of invoice2data (before we make modifications)

alexis-via committed:
Avoid a potential crash

rafaelbn commented:
Hi, thanks for this contribution. Of course this is World-Wide interest. I would like to test ii, could any one add dependencies please?

Runbot fails here:

```
2016-03-03 00:09:11     CRITICAL    server  openerp.service.server:942 preload_registries

Failed to initialize database `3142975-122-5e7e44-all`.
Traceback (most recent call last):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142975-122-5e7e44/openerp/service/server.py", line 929, in preload_registries
    registry = RegistryManager.new(dbname, update_module=update_module)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142975-122-5e7e44/openerp/modules/registry.py", line 370, in new
    openerp.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142975-122-5e7e44/openerp/modules/loading.py", line 319, in load_modules
    modobj.button_install(cr, SUPERUSER_ID, ids)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142975-122-5e7e44/openerp/api.py", line 268, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142975-122-5e7e44/openerp/addons/base/module/module.py", line 414, in button_install
    self.state_update(cr, uid, ids, 'to install', ['uninstalled'], context=context)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142975-122-5e7e44/openerp/api.py", line 268, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142975-122-5e7e44/openerp/api.py", line 399, in old_api
    result = method(recs, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142975-122-5e7e44/openerp/addons/base/module/module.py", line 405, in state_update
    self.check_external_dependencies(module.name, newstate)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142975-122-5e7e44/openerp/addons/base/module/module.py", line 378, in check_external_dependencies
    raise orm.except_orm(_('Error'), msg % (module_name, e.args[0]))
except_orm: ('Error', u'Unable to install module "account_invoice_import" because an external dependency is not met: No module named invoice2data')
```

alexis-via commented:
@rafaelbn I updated .travis.yml to add the invoice2data lib... but I don't know which file I should update for runbot...

alexis-via commented:
It seems that there is a pb in Travis environment: it doesn't have wkhtmltopdf, cf travis logs:

```
File "/home/travis/odoo-8.0/addons/report/models/report.py", line 54, in _get_wkhtmltopdf_bin
raise IOError
IOError
```

alexis-via committed:
Add wkhtmltopdf in travis config

alexis-via committed:
Disable test of invoice2data, because this lib is difficult to install

alexis-via committed:
Fix PEP8

alexis-via committed:
Generate proper PDF metadata according to ZUGFeRD standard

pedrobaeza commented:
@alexis-via, I see that you already figure out how to add wkhtmltopdf.

I have added the dependencies on runbot and launch a rebuild. Let's see if all goes well.

Can you please squash commits a bit, because there are lots of very tiny commits.

alexis-via committed:
Add ZUGFeRD extension schema and tags in XMP metadata

alexis-via committed:
ZUGFeRD invoices are now valid according to this ZUGFeRD validation portal: https://www.din-zugferd-validation.org

alexis-via committed:
Replace pdfminer by PyPDF2 everywhere

RoelAdriaans commented on account_invoice_import/README.rst:
```diff
@@ -0,0 +1,140 @@
+
+.. code::
+
+  sudo add-get install imagemagick tesseract-ocr
```
Minor typo, should be apt-get install <...>

alexis-via committed:
Fix error in README spotted by Roel Adriaans

RoelAdriaans commented:
Testing this module, looks very interesting! We are also interested in porting and using this module with UBL invoices, because that seems to be the default in the Netherlands.

Found a usability issue: In Import Supplier Invoices, the Partners tree forces me to create a new partner, I cannot select current partner. 
This is now a 2 step process, first create a account.invoice.import.config, then go to a partner and select this import config.
(Tested on OCA runbot and local machine with latest odoo 8.0 code)

RoelAdriaans commented:
Runbot error when printing an customer invoice:

```
Traceback (most recent call last):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3143772-122-937a9d/openerp/addons/report/controllers/main.py", line 116, in report_download
    response = self.report_routes(reportname, docids=docids, converter='pdf')
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3143772-122-937a9d/openerp/http.py", line 405, in response_wrap
    response = f(*args, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3143772-122-937a9d/openerp/addons/report/controllers/main.py", line 65, in report_routes
    pdf = report_obj.get_pdf(cr, uid, docids, reportname, data=options_data, context=context)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3143772-122-937a9d/openerp/api.py", line 268, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3143772-122-937a9d/openerp/addons/account_invoice_zugferd/models/report.py", line 28, in get_pdf
    pdf_content)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3143772-122-937a9d/openerp/api.py", line 266, in wrapper
    return new_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3143772-122-937a9d/openerp/addons/account_invoice_zugferd/models/account_invoice.py", line 746, in regular_pdf_invoice_to_zugferd_invoice
    new_pdf_filestream.appendPagesFromReader(original_pdf)
AttributeError: 'PdfFileWriter' object has no attribute 'appendPagesFromReader'
```

alexis-via commented:
@RoelAdriaans-B-informed Could you send me a sample UBL invoice ? In the Netherlands, is it an XML file embedded in the PDF ? Or do you sent the XML file directly (which would seem a bit strange).

About the usability issue: I must say that I don't use the menu entry "Accounting > Config > Import supplier invoices" to create new configurations, only to view or edit them. When I create a new configuration, I go to the partner form, and then I create a new configuration via the M2O field ; it works well.

About the runbot error: the runbot environnement needs to have the addtionnal libs, in particular PyPDF2 and invoice2data... and I don't know how to do that (I know how to add a lib to Travis, I don't know how to add one to the runbot).

alexis-via commented:
@RoelAdriaans-B-informed I'll add a module account_invoice_import_ubl
I fould an example UBL invoice here:
http://docs.oasis-open.org/ubl/os-UBL-2.1/xml/UBL-Invoice-2.1-Example.xml
But it doesn't use real data (for example, there is no VAT number in this invoice), so I would prefer to have a real-life UBL invoice. If you have one, could you send it to me by email ? I'll keep it confidential.

RoelAdriaans commented:
@alexis-via Email with UBL invoices is send. 

Public test invoices: [XML file](http://www.softwarepakket.nl/upload/ublketentest/voorbeelden/referentiefactuur/UBLKetentest_Referentiefactuur_20150102_UBL21.xml) and [PDF Invoice with included xml](http://www.softwarepakket.nl/upload/ublketentest/voorbeelden/referentiefactuur/UBLKetentest_Referentiefactuur_20150102.pdf)

alexis-via committed:
Add module account_invoice_import_ubl

Update README and fix typo

alexis-via commented:
The next step is to rename base_zugferd to a new name without "zugferd"... but, before that, I need to figure out what is common between CII and UBL in term of nomenclature, and what is the underlying standard that specify this thing (because it seems there is one !).

alexis-via commented:
All the standard code for taxes, unit of measure and payment code seems to come from UNECE http://www.unece.org/cefact and be named United Nations Code List (uncl). So I'll propose to move the module base_zugferd to 3 modules:
- product_uncl (depend on product) ; add codes for unit of measure
- account_uncl (depend on account) ; add codes on taxes
- account_payment_uncl (depend on account_banking_payment_export) ; add code on payment types
  that would be hosted in community_data_files.
  If you don't agree with this approach, please yell now... because I plan to do the work very soon.

alexis-via commented:
After more investigation, I decided that the new name for the modules will be:
- product_unece
- account_unece
- account_payment_unece

alexis-via committed:
Adapt code to new UNECE modules from OCA/community-data-files

alexis-via commented:
Here are the UNECE modules: https://github.com/OCA/community-data-files/pull/14
This PR now depends on the PR of the UNECE modules in community-data-files.

alexis-via committed:
Add support for UBL refunds

If you have an UBL refund that I could use to test, please contact me !

alexis-via committed:
FIX order of XML tags when the product has a sale_description

alexis-via committed:
Use base_vat_sanitized

Convert YAML test to unittest

alexis-via committed:
Add support for import of e-fff invoices (Belgian standard)

alexis-via committed:
Add an e-fff invoice to the test suite (kindly provided by Luc de Meyer from Noviat)

Convert tests from YAML to unittest

alexis-via committed:
Add support for local invoice template for invoice2data, see README

alexis-via committed:
Small usability improvement

alexis-via commented:
@RoelAdriaans-B-informed About your remark "Found a usability issue: In Import Supplier Invoices, the Partners tree forces me to create a new partner, I cannot select current partner.". I don't think it's technically possible to select an existing partner because it is a one2many field (if you know a technical solution for that, please tell me). So I decided to set the field as readonly to avoid this frustration.

alexis-via commented:
I got the information that the French standard for e-invoices in PDF with embedded XML will not be ready before september 2016 at least. So there won't be any revolution in the code base in the short term to factorise the code between zugferd and the french standard (which should be very close). So I think that, when the french standard will be released, we don't change the organisation of the modules for v8, but I will do that when I'll port it to Odoo v9 or v10.

So I think that this PR is ready to merge now.

alexis-via committed:
Add note about e-fff invoices in README.

rafaelbn commented:
Is there any possibility to test it functionally?

alexis-via committed:
PEP8 fix

alexis-via commented:
@rafaelbn Yes, sure. The modules are provided with demo data and the modules have unit tests with real invoices that are used by the unit tests, so you can also use these invoices to make your functional tests.

alexis-via committed:
Add partner-contact in oca_dependencies.txt

alexis-via commented:
I'm trying to put travis back to green, but it seems test unit-tests fails on the use of the invoice2data lib... where as, when I run the unit-tests on my laptop, it works perfectly well. I guess there is a problem in the installation of the invoice2data lib via travis, or something around that.

eltanno commented:
Does this work with Odoo 9?

alexis-via commented:
@eltanno this PR is for Odoo 8. You would have to port the modules to odoo 9.

rafaelbn commented:
I cannot test it in runbot. Maybe related to:

```
2016-06-23 12:09:29     CRITICAL    server  openerp.service.server:949 preload_registries

Failed to initialize database `3149310-122-1aef45-base`.
Traceback (most recent call last):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3149310-122-1aef45/openerp/service/server.py", line 936, in preload_registries
    registry = RegistryManager.new(dbname, update_module=update_module)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3149310-122-1aef45/openerp/modules/registry.py", line 370, in new
    openerp.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3149310-122-1aef45/openerp/modules/loading.py", line 311, in load_modules
    modobj.update_list(cr, SUPERUSER_ID)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3149310-122-1aef45/openerp/api.py", line 268, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3149310-122-1aef45/openerp/addons/base/module/module.py", line 651, in update_list
    handler.load_addons()
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3149310-122-1aef45/openerp/http.py", line 1316, in load_addons
    m = __import__('openerp.addons.' + module)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3149310-122-1aef45/openerp/modules/module.py", line 80, in load_module
    mod = imp.load_module('openerp.addons.' + module_part, f, path, descr)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3149310-122-1aef45/openerp/addons/product_multi_image/__init__.py", line 8, in <module>
    from .hooks import pre_init_hook
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3149310-122-1aef45/openerp/addons/product_multi_image/hooks.py", line 5, in <module>
    from openerp.addons.base_multi_image.hooks import pre_init_hook_for_submodules
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3149310-122-1aef45/openerp/modules/module.py", line 79, in load_module
    f, path, descr = imp.find_module(module_part, ad_paths)
ImportError: No module named base_multi_image
```

alexis-via committed:
Better error message

alexis-via committed:
Add debug message

alexis-via committed:
Checks that the directory really exists

alexis-via committed:
More logs

alexis-via committed:
Get invoice2data from pypi, because it is up-to-date

alexis-via committed:
Still trying to get more details about the failure in travis

alexis-via committed:
More logs

alexis-via committed:
Found the problem with travis: invoice2data needs pdftotext

alexis-via committed:
Try to fix travis config

alexis-via committed:
Move download source to akretion with single file (no need for the complete archive), for faster download

alexis-via committed:
Add chmod 755

alexis-via commented:
At last, Travis is OK for this PR now ! Travis is red now only because of account_invoice_merge_purchase which seem to reveal a bug in the official purchase module... but a fix has been made 4 days ago https://github.com/odoo/odoo/commit/830fc8922d32d751c2090940b8217e6628682007 so it seems that travis is not using the latest build of odoo !

Now that Travis is OK for this PR, we can merge it !

alexis-via commented:
Can someone add OCA/partner-contact for runbot ?

pedrobaeza commented:
Runbot dependency added and rebuild launched.

I alerted about the problem in Odoo working on https://github.com/OCA/account-invoicing/pull/160, but I don't know why Travis still complains.

alexis-via commented:
@pedrobaeza Sorry, we also need OCA/community-data-files on runbot (in addition to OCA/partner-contact)

pedrobaeza commented:
No problem, done

alexis-via commented:
@pedrobaeza  There are 2 things left to add in runbot: the invoice2data + PyPDF2 lib and pdftotext binary in version 3.04, cf travis.yml

rvalyi commented:
@alexis-via @pedrobaeza please hold this merge a bit so we can review it. We have a similar electronic invoice import interest in Brazil. The XML format is different but the overall logic might be the same. I would like to see if we could suggest easy design changes so we could swap the XML mapping logic by our own. Regards. cc @renatonlima

alexis-via commented:
@rvalyi Could you send me 1 or 2 brasilian XML invoices, so that I can have a look ?

alexis-via committed:
FIX: avoid a crash on draft invoices without due date

alexis-via commented:
Travis is red, but it's unrelated ; it is still the issue with account_invoice_merge_purchase

@rvalyi  @renatonlima  I didn't get any news from you about Brazilian XML invoices, and I don't want to hold this merge any longer because this PR is already 7 months old. If we want to make some design change, we can do it later, or do it when I'll port these modules to v9 or v10. What do you think ? Or could you post here the details  of what you would like to change ?

gurneyalex commented:
:+1:

RoelAdriaans commented:
:+1: 

Did functional tests with this code, no problems!

alexis-via committed:
Extract regular PDF invoice import from account_invoice_import to a dedicated module account_invoice_import_invoice2data

Update README.rst and headers to latest OCA conventions.

alexis-via committed:
Adapt code to plug the module 'ovh_supplier_invoice'

Better key names in the parsed_inv dict
parsed_inv doesn't need to be JSON serializable anymore (small drawback: the invoice is parsed a second time on the second step... but the second step is rarely used)

alexis-via committed:
Fix product matching

Move code from account_invoice_import_invoice2data to account_invoice_import

alexis-via committed:
FIX PEP8 and account_invoice_import_invoice2data unittests

Update REAME and some interface strings about UBL being an ISO standard
Small code changes

alexis-via commented:
FYI, I'm starting to work on the import of UBL sale orders (modules sale_order_import +  sale_order_import_ubl). For that, I want to reuse some methods of account_invoice_import (the methods that match partners, products, currencies, etc...) and I think that sale_order_import should NOT depend on account_invoice_import. So I need to have a new base module that would be common for account_invoice_import and sale_order_import. I propose to name that module "base_business_document_import". If you have a suggestion for a better module name, please propose it now !

alexis-via committed:
PEP8 fix

pedrobaeza commented:
Alexis, as `sale` module depends on `account`, I don't see special problem to keep that dependency.

rvalyi commented:
@alexis-via sorry for the delay. Here is an example of such a Brazilian electronic invoice:
https://gist.github.com/rvalyi/a62e892510ebb885f00b05413de71d4e
in the structure, you have information about the origin company in the "emit" section, information about the destination company in the "dest" section, In this example you have only 1 product line in the "det" section. ICMS is our local VAT, but you often have 3 or 4 other kind of taxes per lines (you have the other ones in the "total" section. We can also have carrier information in the "transp" section and information about the payment parcels in the "pag" section.

We are analysing with @renatonlima to give you a more comprehensive feedback about the code of account_invoice_import. Basically we will need a few hook extracted that we will suggest.

Also We think it would be cool if your module could deal with exception with that new module we just contributed: https://github.com/OCA/account-invoicing/pull/173
In Brazil we have so much tax complexity that just like when we import orders from an ecommerce, we will need a sophisticated management of all the possible imporation error cases (we have many more entities that can be badly imported, like all the kind of taxes) where manual action will need to be taken.

rvalyi commented:
@alexis-via I'm still analyzing but soon I promise a better feedback.

## What I can say already:

in your module when you find an error, you raise an exception straight. That's fine for your use case. However, in our case, as I said, the fiscal data is so complex that instead of a hard fail for any error we will often prefer logging the error using the account_invoice_exception WIP module and finish the importation and then ask for a manual correction of the exceptions.

Of course I can understand you don't want to bloat your module with a dependency upon account_invoice_exception that is very fine. Instead I suggest that everywhere you are raising a UserError during the line importation, you instead call some 2 extracted methods like:

```
def process_invoice_exception(self, field_name, message, invoice_vals):
   raise UserError(message)
end

def process_line_exception(self, field_name, message, invoice_line_vals):
   raise UserError(message)
end
```

and instead of raising UserError directly you will call them with something like:

```
invoice_vals = self.process_line_exception(field_name, message, invoice_vals)
invoice_line_vals = self.process_line_exception(field_name, message, invoice_line_vals)
```

If you do that, in our Brazilian module, we will override these process(_line)_exception methods and we could can an Exception entity (in the invoice_line_vals) instead and keep importing. I'll confirm that but I think we can hard block on account.invoice errors and eventually keep importing for account.invoice.line errors (but you will always hard block in your own module).

## Open issues at this stage:

we will need a few hooks extracted in the **_prepare_create_invoice_vals** method, specially to deal with:
- in v8, we will need a way to inject a context in the on_change_partner_id and on_change_product_id because this is where we tunnel the fiscal parameters we need to select taxes and other fiscal crap in v8.
- a hook deal with the due_date: in Brazil we nearly always have partial payments which are duly listed inside our electronic invoice XML
- a hook to deal with banks and IBAN as we don't have an iban here.

That is all I see at this stage but we should still check. We are available on hangout anytime if you want to talk about that.

alexis-via commented:
@pedrobaeza I think you didn't get my point ; the module account_invoice_import contains a wizard and a new object (account.invoice.import.config). These things are not needed for the module sale_order_import, so I don't want sale_order_import to depend on account_invoice_import. For that, I need to move some code of account_invoice_import to a new module "base_business_document_import".

My aim is to have the following setip:
- "account_invoice_import" should depend on "account" and "base_business_document_import"
- "sale_order_import" should depend on "sale" and "base_business_document_import" (and NOT on "account_invoice_import")

pedrobaeza commented:
OK, got it. Better call it `base_document_import` to be shorter, but meaningful.

alexis-via commented:
Well, "base_document_import" make me think of the "document" module... which is completely unrelated. I used the term "business document" because this term is used a lot in the UBL standard, cf http://ubl.xml.org/

If we really want to make the name shorter, we could name it "business_document_import" without the "base" prefix... but I like the "base" prefix for some technical modules that have common/shared features.

pedrobaeza commented:
OK, let's go with the larger one

alexis-via committed:
Add module base_business_document_import and start to move some code to it.

alexis-via committed:
Better/cleaner pivot invoice dict format (no more IDs in the pivot invoice dict, cleaner organisation)

Code refactoring: move code in base_business_document_import, factorise code for tax matching (it was duplicated in UBL and ZUGFeRD)
Now support PDF with embedded UBL XML file
Enable unittests on account_invoice_import_ubl
More absolute xpath in account_invoice_import_ubl instead of relative xpath

WARNING: these are big changes, I may have broken a few details

alexis-via committed:
PEP8 fix

alexis-via commented:
@rvalyi I have almost finished my big changes in this PR, and I propose to take into account the changes you want after I finish those big changes.

rvalyi commented:
@alexis-via no problem. In any case I think only minor changes will be required for us to be able to reuse your module here in Brazil.

alexis-via committed:
Add module base_ubl (common methods to generate and parse UBL files)

Adapt module account_invoice_import_ubl to use the new base_ubl module
Small fixes

alexis-via committed:
Parse customer/supplier ref in UBL and allow to match partner on 'ref'

Better generation of address block in UBL (make it coherent with the datamodel of Odoo)
Add generation of several UBL blocks: language, delivery, payment terms, customer party, spplier party
Add parsing of zip in UBL party (will be used in the future for delivery partner match)

alexis-via committed:
PEP8 fix

alexis-via committed:
Add matching of delivery partner

Use country code and state code to match partners
UBL: Add delivery terms and line item UBL XML block generation
UBL: add parsing of delivery block

alexis-via committed:
Add module base_phone_business_document_import

Add unitests in base_business_document_import
Small code enhancements/simplifications

alexis-via committed:
Add unitest on partner matching and update code of partner matching

alexis-via committed:
FIX make sure prices are without taxes in ZUGFERD XML file

alexis-via committed:
Add a unittest on _match_uom

Code cleanup/minor changes

alexis-via committed:
Add module base_business_document_import_stock

Rename key 'quantity' to 'qty' in all parsing dicts['lines']
Add common methods compare_lines() and post_create_or_update()
Make sure price_unit is always untaxed in UBL XML files

alexis-via committed:
Update invoice lines on "Update Existing" invoice import

Add button "Update from File" on supplier invoice form
Use float_round() instead of round()

alexis-via committed:
Fix crash

alexis-via committed:
Convert YAML tests to unittest

Standardize tests/files as the subdir for test import files

alexis-via committed:
Rename base_phone_business_document_import to base_business_document_import_phone for consistency

Add connector-telephony in oca_dependencies.txt

alexis-via committed:
Add phonenumbers lib in travis config

alexis-via committed:
import phonenumbers in try/except

alexis-via committed:
Add UBL 2.1 XSD files (can be used by all UBL modules ; avoids duplication of UBL XSD files)

alexis-via committed:
Fix code + PEP8

alexis-via committed:
Add and update README.rst

alexis-via committed:
Use StreetName/AdditionalStreetName instead of AddressLine blocks, because only the first one is accepted by Chorus Factures for example.

alexis-via committed:
Update UBL methods to generate lines

Add UBL methods to generate some tax blocks

alexis-via committed:
Add modules account_invoice_ubl + base_ubl_payment

alexis-via committed:
Add partner identification hook in UBL XML party block generation

Print numbers with the approriate number of digits in the XML

alexis-via committed:
Move payment_means_49 to account_payment_unece

alexis-via committed:
Add multi-company ir.rule on account.invoice.import.config

alexis-via committed:
Use vat_sanitiezd in base_ubl

alexis-via committed:
Add support for UBL 2.0 (required for e-fff)

alexis-via committed:
Update README

alexis-via committed:
Handle lang in UBL XML file generation

alexis-via committed:
FIX Don't use _inherit = ['business.document.import']

because we want to have access to the code of the modules that inherit business.document.import

alexis-via committed:
account_invoice_import_invoice2data: allow to use both a local template dir and the built-in templates

Also allow to use only a local template dir
README updated to explain how to configure all this

alexis-via committed:
Transfer log level of Odoo to invoice2data lib

Special thanks to Sébastien Beau for his help to achieve this

alexis-via committed:
Fix method call

alexis-via committed:
Allow import of UBL invoices without InvoicedQuantity tag in invoice lines

rousseldenis commented on account_invoice_import_invoice2data/wizard/account_invoice_import.py:
```diff
@@ -0,0 +1,90 @@
+            'invoice2data_templates_dir', False)
+        logger.debug(
+            'invoice2data local_templates_dir=%s', local_templates_dir)
+        templates = None
```
Should be templates = []

rousseldenis commented:
@alexis-via It seems that Zugferd validation does not pass on data (Structure is ok).
Number of decimals is 2 except for rebate (4).
https://www.din-zugferd-validation.org

Should we cast data ?

alexis-via commented:
Moved to edi project.
